src/devtools/client/webconsole/utils/autocomplete.ts

```diff
@@ -0,0 +1,249 @@
+  const expression = input.slice(0, input.lastIndexOf("[") + 1);
+  return isMemberExpression(expression + `"a"]`);
+}
+function isDotNotation(input: string) {
+  return isMemberExpression(input + `a`);
+}
+function isAccessingObjProperty(input: string) {
+  return isBracketNotation(input) || isDotNotation(input);
+}
+
+function getPropertiesForObject(object: ObjectFront): string[] {
+  const preview = object.preview;
+  const properties = [];
+
+  // The object preview may be undefined until the user expands the object
+  // in the scopes panel. This applies primarily to prototype objects.
+  if (preview) {
+    const objectProperties = preview.properties.map(b => b.name);
+    properties.unshift(...objectProperties);
+  } else {
+    if (["Object", "Array"].includes(object.className)) {
+      const prototypeProperties = Object.getOwnPropertyNames(Object.prototype);
+      properties.unshift(...prototypeProperties);
+    }
+
+    if (object.className === "Array") {
+      const prototypeProperties = Object.getOwnPropertyNames(Array.prototype);
+      properties.unshift(...prototypeProperties);
+    }
+  }
+
+  const prototype = preview?.prototypeValue;
+
+  // Recursively gather the properties through the prototype chain.
+  if (prototype?._object) {
+    const prototypeProperties = getPropertiesForObject(prototype._object);
+    properties.unshift(...prototypeProperties);
+  }
+
+  return properties;
+}
+
+export function getParentExpression(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(0, input.lastIndexOf("["));
+  } else {
+    return input.slice(0, input.lastIndexOf("."));
+  }
+}
+
+export function shouldPropertyAutocomplete(input: string) {
+  if (!isAccessingObjProperty(input)) {
+    return false;
+  }
+
+  const parentExpression = getParentExpression(input);
+
+  if (!parentExpression) {
+    return false;
+  }
+
+  // Right now, this only works for direct property access (e.g., a.b)
+  // and doesn't support chained expressions (e.g., a.b.c.d). This is because
+  // we don't (yet) have the chained values in the fetched frame scope data.
+  return isIdentifier(parentExpression);
+}
+function shouldVariableAutocomplete(input: string) {
+  return isIdentifier(input);
+}
+
+function getBinding(name: string, frameScope: FrameScope) {
+  return frameScope.scope.bindings.find(b => b.name === name);
+}
+function getBindingNames(scope: FrameScope): string[] {
+  if (!scope?.scope) {
+    return [];
+  }
+
+  return scope.scope.bindings.map(b => b.name);
+}
+function getGlobalVariables(scopes: Scope) {
+  const variableNames = [];
+  const globalObject = scopes.parent.object;
+
+  if (globalObject) {
+    const properties = getPropertiesForObject(globalObject._object);
+    variableNames.push(...properties);
+  }
+
+  return variableNames;
+}
+
+export function getLastToken(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(input.lastIndexOf("[") + 1);
```

Comment:
Yeah this is a repetition of the logic above, feels like there is a clean abstraction in here somewhere?

Reply:
Yeah. Can't think of a way to simplify this right now though, I'll sleep on it.

Reply:
I would probably have a tiny data structure that looked like:

```
{
  left: "someObject",
  right: "somePartiallyWrittenPropertyName",
  separator: "."
}
```
Or you could do `parent/child`, because I think that's another metaphor we use in here somewhere.